src/vs/base/parts/quickinput/browser/quickInput.ts

```diff
@@ -483,19 +484,25 @@ class QuickPick<T extends IQuickPickItem> extends QuickInput implements IQuickPi
 	}
 
 	get value() {
-		return this._value;
+		return this._value ?? '';
 	}
 
 	set value(value: string) {
 		if (this._value !== value) {
+			if (typeof this._value !== 'undefined') {
```

Comment:
If I change this to `if (typeof this._value !== 'undefined' || this.visible) {` then all tests appear to pass... however, I need to get a better understanding I think.

Reply:
The listener on the textbox below in `show()` is only registered while the QuickPick is visible. Calling `this.onDidChangeValueEmitter.fire(this._value)` here might still make sense when the QuickPick is hidden. Or is that specifically what you want to change?

Reply:
@chrmarti the issue is https://github.com/microsoft/vscode/issues/137279 where when you set the value of the input box programmatically, we are not ending up refiltering the picker, because of this logic:
* `this._value = value || '';` is being called [here](https://github.com/microsoft/vscode/blob/26fe37ca3d9dcc934ffcacd82443a38c9b192dd0/src/vs/base/parts/quickinput/browser/quickInput.ts#L491) to update the model of the picker
* this is followed by a call to `update`
* the input box updates its value [here](https://github.com/microsoft/vscode/blob/26fe37ca3d9dcc934ffcacd82443a38c9b192dd0/src/vs/base/parts/quickinput/browser/quickInput.ts#L969)
* the `onDidChange` listener fires from the input box
* however the picker refuses to refilter because it thinks the value has not changed:

https://github.com/microsoft/vscode/blob/26fe37ca3d9dcc934ffcacd82443a38c9b192dd0/src/vs/base/parts/quickinput/browser/quickInput.ts#L737-L739

Reply:
Would Tyler's fix improve if it was wrapped in an `if (this.visible)`? (A similar approach would be to move Tyler's change into `update()` which also updates the filtering when the items update.)

On not checking `value === this.value`: We would want to avoid triggering additional events on the API, but we could filter unconditionally.

Reply:
@chrmarti but I think there really is a bug in your logic how you decide to refilter or not. 

This code ensures to only update the input box when the value differs:

https://github.com/microsoft/vscode/blob/26fe37ca3d9dcc934ffcacd82443a38c9b192dd0/src/vs/base/parts/quickinput/browser/quickInput.ts#L968-L970

But then in the change listener this condition checks on a different value:

https://github.com/microsoft/vscode/blob/26fe37ca3d9dcc934ffcacd82443a38c9b192dd0/src/vs/base/parts/quickinput/browser/quickInput.ts#L737-L739

In other words: we knew the value would change and updated the input box but then in the listener for the change we forget about this fact?

Reply:
The way I think about it: The first propagates the value from the model object to the input box and forgets to filter (the bug you are fixing). The second propagates the value from the input box to the model object and correctly filters.

Refiltering in `update()` (your first code reference, maybe further down in that method after the items were updated) if the value changed seems like it might address this first case.

I haven't debugged into it yet, so I might be missing something.

Reply:
Yeah good point, I have opened https://github.com/microsoft/vscode/pull/137857 now with a different strategy that goes more into that direction.